app/src/main/java/com/stevesoltys/seedvault/transport/backup/PackageService.kt

```diff
@@ -50,17 +53,14 @@ internal class PackageService(
                 }
             }
 
-            val eligibleApps =
-                backupManager.filterAppsEligibleForBackupForUser(myUserId, packages.toTypedArray())
```

Comment:
Why not leave it to the system to tell us which apps are eligible?

Reply:
It returns the wrong list.
https://cs.android.com/android/platform/superproject/+/master:frameworks/base/services/backup/java/com/android/server/backup/UserBackupManagerService.java;l=585;drc=dada247f224cfd53ab6b5a56c2bd136badea12f0;bpv=0;bpt=1 uses OperationType.BACKUP. We want OperationType.MIGRATION

I'm doing the necessary checks in BackupCoordinator instead

Reply:
Maybe another option around this could be to take the scheduling into our own hands and then use the methods that allow us to pass the operation type.

Doing the scheduling ourselves has other advantages, e.g. offline, metered networks, usb, etc. which are already tricky with AOSPs scheduling. But it would be way more work as well.

Reply:
I don't see how those rules get used? Are they not asking the transport anymore what it considers eligible?

How certain are you that there is no other way to call into system methods with a different operation type so we get the OS rules without having to duplicate them ourselves? E.g. can we call `BackupManagerService.getEligibilityRulesForOperation(operationType)`?

Do you know what happens if we try to back up an app that AOSP doesn't consider eligible even when using `OperationType.MIGRATION`?

Reply:
Concerning scheduling, I am unsure that can be done without OS changes.

`getEligibilityRulesForOperation` is not exposed in current public nor system APIs. Without OS modifications we cannot make use of it. The ideal for most compatibility for this feature is to have no OS changes required. A number of calls to determine eligibility are hardcoded in the system.

Concerning app eligibility for backup, the only difference is that `OperationType.BACKUP` only checks `allowBackup` while `OperationType.MIGRATION` also checks first if `allowBackup` can be ignored. So there would be no case where the list of eligible apps returned by `OperationType.BACKUP` contains apps that are not present in  `OperationType.MIGRATION` https://cs.android.com/android/platform/superproject/+/master:frameworks/base/services/backup/java/com/android/server/backup/utils/BackupEligibilityRules.java;l=163;drc=4875b203ceba30d2d50d642da04064ca37b66e43;bpv=0;bpt=1